[libilbc] New port

geraldcombs opened the pull request:
**Describe the pull request**

This adds a port for libilbc, an implementation of the Internet Low Bit Rate Codec (iLBC).

- What does your PR fix? Fixes #14836

- Which triplets are supported/not supported?

Tested with x64-osx, but should work with *-windows and *-linux.

- Have you updated the CI baseline?

Yes.

- Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?

Yes.

geraldcombs committed:
[libilbc] New port

libilbc is an Internet Low Bit Rate Codec (iLBC) / RFC 3951 codec.
Closes #14836.

JackBoosY commented on ports/libilbc/portfile.cmake:
```diff
@@ -0,0 +1,39 @@
+    ARCHIVE ${ARCHIVE}
+    REF ${ILBC_VERSION}
+)
+
+string(COMPARE EQUAL "${VCPKG_LIBRARY_LINKAGE}" "dynamic" BUILD_SHARED_LIBS)
```
```suggestion
```

JackBoosY commented on ports/libilbc/portfile.cmake:
```diff
@@ -0,0 +1,39 @@
+vcpkg_configure_cmake(
+    SOURCE_PATH ${SOURCE_PATH}
+    PREFER_NINJA
+    OPTIONS
+        -DBUILD_SHARED_LIBS=${BUILD_SHARED_LIBS}
```
```suggestion
```
We don't need this, `vcpkg_configure_cmake` already do that.

JackBoosY commented on ports/libilbc/portfile.cmake:
```diff
@@ -0,0 +1,39 @@
+vcpkg_install_cmake()
+file(INSTALL ${SOURCE_PATH}/COPYING DESTINATION ${CURRENT_PACKAGES_DIR}/share/${PORT} RENAME copyright)
+
+file(GLOB EXES "${CURRENT_PACKAGES_DIR}/bin/*.exe")
+if(EXES)
+    file(REMOVE ${EXES})
+endif()
+
+file(GLOB DEBUG_EXES "${CURRENT_PACKAGES_DIR}/debug/bin/*.exe")
+if(DEBUG_EXES)
+    file(REMOVE ${DEBUG_EXES})
+endif()
```
We should use `vcpkg_copy_tools` instead.

JackBoosY commented on ports/libilbc/portfile.cmake:
```diff
@@ -0,0 +1,39 @@
+file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/include")
+file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/share")
+if(VCPKG_LIBRARY_LINKAGE STREQUAL "static")
+    file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/bin" "${CURRENT_PACKAGES_DIR}/debug/bin")
```
With `vcpkg_copy_tools`, I think this should be deleted.

JackBoosY commented on ports/libilbc/portfile.cmake:
```diff
@@ -0,0 +1,39 @@
+        -DBUILD_SHARED_LIBS=${BUILD_SHARED_LIBS}
+)
+vcpkg_install_cmake()
+file(INSTALL ${SOURCE_PATH}/COPYING DESTINATION ${CURRENT_PACKAGES_DIR}/share/${PORT} RENAME copyright)
```
Please move this line to the end of this file.

JackBoosY commented on scripts/ci.baseline.txt:
```diff
@@ -598,6 +598,7 @@ libics:x64-uwp=fail
 libigl:arm64-windows=fail
 libigl:arm-uwp=fail
 libigl:x64-uwp=fail
+libilbc:arm-uwp=fail
```
Does that expected? If it is, please add `Support: !(arm&uwp)` to `CONTROL` file.

geraldcombs replied:
The build fails with

         D:\buildtrees\libilbc\src\3.0.3-bfa5adc7c7.clean\rtc_base/system/arch.h(79,1): fatal error C1189: #error:  Please add support for your architecture in rtc_base/system/arch.h (compiling source file D:\buildtrees\libilbc\src\3.0.3-bfa5adc7c7.clean\modules\audio_coding\codecs\ilbc\decode.c) [D:\buildtrees\libilbc\arm-uwp-dbg\ilbc.vcxproj]

and https://github.com/TimothyGu/libilbc/blob/main/rtc_base/system/arch.h doesn't handle `_M_ARM` so I'm guessing the answer is "yes". I updated `CONTROL` accordingly.

geraldcombs committed:
Update ports/libilbc/portfile.cmake

Co-authored-by: Jack·Boos·Yu <47264268+JackBoosY@users.noreply.github.com>

geraldcombs committed:
Update ports/libilbc/portfile.cmake

Co-authored-by: Jack·Boos·Yu <47264268+JackBoosY@users.noreply.github.com>

geraldcombs committed:
[libilbc] Use vcpkg_copy_tools, update arm-uwp.

JackBoosY reviewed (changes requested):
There are some questions:
1. libilbc installed `libilbc.pc` in _PACKAGES/lib/pkgconfig_, so we should add `vcpkg_fixup_pkgconfig` after `vcpkg_install_cmake`.
2. libilbc's docs is installed in _PACKAGES/share/doc_, the correct path should be _PACKAGES/share/libilbc/doc_.
3. I found the tool is test program, so we shouldn't install it, please add patch or option to disable it.
4. In _PACKAGES/include/ibc_export.h_ line 5:
```cpp
#ifdef ILBC_STATIC_DEFINE
#  define ILBC_EXPORT
#  define ILBC_NO_EXPORT
#else
#  ifndef ILBC_EXPORT
#    ifdef ilbc_EXPORTS
        /* We are building this library */
#      define ILBC_EXPORT __declspec(dllexport)
#    else
        /* We are using this library */
#      define ILBC_EXPORT __declspec(dllimport)
#    endif
#  endif

#  ifndef ILBC_NO_EXPORT
#    define ILBC_NO_EXPORT 
#  endif
#endif
```
The judgment `#ifdef ILBC_STATIC_DEFINE` shoud be `#if 1` when building static, so please modify this line after build (use `vcpkg_replace_string`).

geraldcombs committed:
[libilbc] More packaging cleanups.

Add vcpkg_fixup_pkgconfig after vcpkg_install_cmake.

Install docs into $PREFIX/share/libilbc.

Remove ilbc_test after a dynamic build and the bin and debug/bin dirs
after a static build.

Fixup include/ilbc_export.h for static builds.

geraldcombs commented:
>     1. libilbc installed `libilbc.pc` in _PACKAGES/lib/pkgconfig_, so we should add `vcpkg_fixup_pkgconfig` after `vcpkg_install_cmake`.

Done.

>     2. libilbc's docs is installed in _PACKAGES/share/doc_, the correct path should be _PACKAGES/share/libilbc/doc_.

libilbc's CMakeLists.txt calls `GNUInstallDirs` so this was done via `-DCMAKE_INSTALL_DOCDIR=share/${PORT}`.

>     3. I found the tool is test program, so we shouldn't install it, please add patch or option to disable it.

Done via a patch.

>     4. In _PACKAGES/include/ibc_export.h_ line 5:
>  [ ... ]
> The judgment `#ifdef ILBC_STATIC_DEFINE` shoud be `#if 1` when building static, so please modify this line after build (use `vcpkg_replace_string`).

Done.

vicroms commented:
Thanks for the PR!
